### ISSUE TITLE
Mojo SDK: Provide an abstract TaskRunner interface

### ISSUE THREAD

**Issue by j9brown** [domokit/mojo#645]
Background: We'd like to be able to publish "public" code to help clients use various Mojo interfaces (eg. audio, graphics, etc.) without forcing a dependency on //base but the Mojo SDK doesn't have everything we need yet.

This is a feature request for some kind of abstract "TaskRunner" or "Looper" interface so that code in the Mojo SDK can post tasks to the run loop without needing to know whether the application is implemented using base::MessageLoop, mojo::RunLoop, or something else.  (The caller would be responsible for providing the correct implementation.)

See also the implementation of ApplicationImpl::Terminate which may also benefit from a message loop abstraction.

**Comment by j9brown** [comment]
cc @johngro 

See circular_buffer_media_pipe_adapter.h for an example of code in the Mojo SDK which needs a message loop abstraction.  The code is currently coupled directly to mojo::RunLoop which makes it impossible to use in an application derived from base::MessageLoop.

**Comment by jamesr** [comment]
We should do this.  It's tricky to do well.  The best place to start would be to define a good Task abstraction and the ownership model around that then go from there. std::function is a poor fit for many of our uses as it requires everything it captures be copyable (regardless of how the std::function is used) and we make extensive use of move-only types.  A hypothetical future std::unique_function<> would be nice but it doesn't exist in the standard yet.  Once a Task is in place the next tricky bit is figuring out ownership and shutdown of the runner, but there are a few choices we can make there if we're ok depending on std::shared_ptr+weak_ptr to help.

We won't be able to provide something maximally efficient and general since we'll have to make some choices in the name of being general that make certain use cases less efficient, so any APIs that rely on this abstraction should be written with customized implementations of the contract in mind.

**Comment by viettrungluu-cr** [comment]
It's pretty hard to provide a task runner abstraction without allowing the
tasks to own things. (For the EDK's purposes, I can get away with
std::function, since I just require that lifetime of task runners to be
such that they execute all tasks posted to them, but this isn't appropriate
for general task runners.)

With C++14 lambdas, it'd be pretty easy to come up with something
std::unique_function-ish, since you can capture move-only things. With just
C++11, we'd also need to come up with something base::Bind-ish (which would
be an additional layer of annoyance/difficulty).

(It's worth noting that base::Callback is hardly ideal, but it works well
enough. We could come up with something base::Bind-ish that produces
std::functions, and that'd work well enough also, though the semantics
would be odd with move-only things, but only in the same way that
base::Callback is.)

**Comment by j9brown** [comment]
With C++11 we can circumvent the problem by wrapping move-only owned objects with std::shared_ptr.  Gross but perhaps adequate in the short term.

**Comment by jamesr** [comment]
I don't think we would want to wrap every move-only type in std::shared_ptr in code using the SDK.  If we want to write reusable code that uses std::function we'd probably be better off unwrapping to copyable types explicitly (i.e. grab raw pointer out of std::unique_ptr<>, raw handles out of handle wrappers) and re-wrapping on the receiving side but that's still pretty gross.

**Comment by j9brown** [comment]
Well in C++14 we won't have to worry about this.  So all we need is a
stopgap for now.  I'll leave it to you folks to decide what the best
interim approach is.

(When do we think we could start using C++14?)

**Comment by jamesr** [comment]
C++14 doesn't really help all that much either.  C++14 lets you construct a lambda with capture expressions which a lambda to capture move-only types.  That's useful for many things but in order to transfer the lambda itself to a generic task scheduling API you need to perform type erasure on the lambda as the lambda is an unutterable type that the generic API cannot know.  C++{11,14,17,...} provide only std::function as a type-erased dispatchable object.  std::function requires that the thing it type-erases must be copyable (even if the std::function itself is never copied) so you can't take a lambda that captures a move-only type and wrap it in an std::function.

std::unique_function, which doesn't exist yet but is likely to be proposed for a future version of the C++ standard, is a type that performs type-erasure on a dispatchable object that can be moved but not necessarily copied. It could hold on to a C++14 lambda that captures move-only types and potentially solve all of our problems, except that it doesn't actually exist anywhere yet.  We could implement this type ourselves it's just work.

**Comment by j9brown** [comment]
Ahh I see.  Tricky!

**Comment by viettrungluu-cr** [comment]
The problem with unwrapping/rewrapping is that it requires a very strict
lifetime model -- in particular, posted tasks *must* be executed
(otherwise, you get leaks).

Maybe we're willing to live with this (what do you do if you shut down with
pending tasks? check-fail?), but it's a bit tough to swallow for a general
facility.

**Comment by viettrungluu-cr** [comment]
Note that strictly speaking, the fact that std::function itself is copyable
isn't the major problem (e.g., base::Callback is copyable, but can deal
with capturing move-only types).

C++14 helps in that it means that we don't have to come up with a move-only
bind (which we'd need to do with C++11).

As you say, even with C++14 we'd still either have to come up with 1)
something like std::unique_function, 2) something like std::function that
can wrap a move-only lambda (possibly with undefined semantics if anyone
tries to execute it twice -- like base::Callback), or 3) something that can
transform a move-only lambda into something that std::function can wrap.

3) is super-easy, especially if we're willing to require some explicitness.
I don't think 1) or 2) are especially hard either, depending on how
efficient you want the implementation to be.
